### PR TITLE
Adjoints ex 6

### PR DESCRIPTION
**Pull request by vikramvgarg**
Adjoint example 6 showing the use of Adjoint Refinement Estimator with Flux QoIs (Adjoint Dirichlet Boundary Conditions).

### CODE REVIEW

**Comment by jwpeterson** [comment]
Not really sure what could possibly be going on here. How do you have conflicts with files you haven't even modified? Also, I wish you would not have closed #1350, as we have now lost the discussion we were having over there...

**Comment by vikramvgarg** [comment]
To obtain the operator_spacing.pl file, I decided to rebase off the latest master. When I tried git rebase master, I got an error saying I cant rebase because my branch had diverged from vikram/adjoints_ex_6. 

I needed to do a git rebase origin/master , which I eventually did. But I did a git pull before that which messed things up. 

Not sure how to untangle this.

**Comment by jwpeterson** [comment]
> Not sure how to untangle this.

You need to have a remote for the upstream repo, if you don't already:
```
git remote add upstream git@github.com:libMesh/libmesh.git
```
or you can use the https URL if you prefer.
Then
```
git fetch upstream
```
and finally, with your branch checked out:
```
git rebase upstream/master
```
and fix the conflicts, if there are any. I suspect the conflicts are coming from the fact that your origin's master and the upstream master are out of sync, but it could be some other problem. Finally, you will need to force push to the `adjoints_ex_6` branch in your fork. Since you haven't changed the branch name, you could probably also reopen the old PR (#1350) and close this messed up one.

**Comment by jwpeterson** [comment]
I just tested it, and to fix the conflicts in your case, during the rebase step just do:
```
git rebase -Xours upstream/master 
```

**Comment by jwpeterson** [comment]
> But I did a git pull before that which messed things up.

IMO `git pull` is one of the worst commands for new users to run, as it typically puts them in a state they either don't understand, or isn't optimal for making pull requests. I'd avoid typing that unless you are sure you know what it's going to do. Better to learn how the individual `fetch`, `merge`, and `rebase` commands do before trying to use them together...

**Comment by vikramvgarg** [comment]
Cool thanks. Let me try rebasing with the options you have stated.

Also, staying off git pull from now on.

**Comment by vikramvgarg** [comment]
This is the result of git rebase -Xours origin/master (upstream in my machine is reserved for GRINS) for me. Does it look safe to push ?

$ git rebase -Xours origin/master
First, rewinding head to replay your work on top of it...
[detached HEAD c7f7dc1] New example that illustrates use of adjoint refinement estimator in conjunction with adjoint dirichlet boundary conditions
 Committer: Vikram Garg <vikram@scott.ices.utexas.edu>
Your name and email address were configured automatically based
on your username and hostname. Please check that they are accurate.
You can suppress this message by setting them explicitly:

    git config --global user.name "Your Name"
    git config --global user.email you@example.com

After doing this, you may fix the identity used for this commit with:

    git commit --amend --reset-author

 11 files changed, 1840 insertions(+)
 create mode 100644 examples/adjoints/adjoints_ex6/Makefile.am
 create mode 100644 examples/adjoints/adjoints_ex6/adjoints_ex4.C
 create mode 100644 examples/adjoints/adjoints_ex6/element_postprocess.C
 create mode 100644 examples/adjoints/adjoints_ex6/femparameters.C
 create mode 100644 examples/adjoints/adjoints_ex6/femparameters.h
 create mode 100644 examples/adjoints/adjoints_ex6/general.in
 create mode 100644 examples/adjoints/adjoints_ex6/poisson.C
 create mode 100644 examples/adjoints/adjoints_ex6/poisson.h
 create mode 100644 examples/adjoints/adjoints_ex6/poisson.in
 create mode 100755 examples/adjoints/adjoints_ex6/run.sh
Committed: 0001 New example that illustrates use of adjoint refinement estimator in conjunction with adjoint dirichlet boundary conditions
Auto-merging examples/adjoints/adjoints_ex6/adjoints_ex6.C
[detached HEAD f8c4767] Modified configure.ac to account for new adjoint example 6
 Committer: Vikram Garg <vikram@scott.ices.utexas.edu>
Your name and email address were configured automatically based
on your username and hostname. Please check that they are accurate.
You can suppress this message by setting them explicitly:

    git config --global user.name "Your Name"
    git config --global user.email you@example.com

After doing this, you may fix the identity used for this commit with:

    git commit --amend --reset-author

 11 files changed, 2418 insertions(+), 900 deletions(-)
 create mode 100644 examples/adjoints/adjoints_ex6/Makefile.in
 rename examples/adjoints/adjoints_ex6/{adjoints_ex4.C => adjoints_ex6.C} (94%)
 rewrite examples/adjoints/adjoints_ex6/femparameters.C (85%)
 rewrite examples/adjoints/adjoints_ex6/femparameters.h (88%)
Committed: 0002 Modified configure.ac to account for new adjoint example 6
All done.

**Comment by jwpeterson** [comment]
It looks like it worked, but please, do what it says and add a ~/.gitconfig file with your name and email address like below so that your commit metadata is always correct.
```
[user]
  name = Vikram Garg
  email = vikram.v.garg@gmail.com
```

Then fix the existing commits by doing:
```
git commit --amend --author="Vikram Garg <vikram.v.garg@gmail.com>"
```

**Comment by jwpeterson** [comment]
Also your history is still showing two commits when it should only show one (your first one still contains incorrectly named files, formatting errors, etc.) Ordinarily we would request that bootstrap-generated changes and actual code should be in separate commits, but splitting it up at this point is not going to be worth the trouble.

**Comment by vikramvgarg** [comment]
It failed make check because there seem to be changes in how Number VectorValue is to be used. Fixing this now.

**Comment by jwpeterson** [comment]
It failed to compile when libmesh was configured with `--enable-complex`. The right fix might be to use `RealVectorValue` instead of `NumberVectorValue`.

```
../../../../examples/adjoints/adjoints_ex6/element_postprocess.C: In member function 'virtual void PoissonSystem::element_postprocess(libMesh::DiffContext&)':
../../../../examples/adjoints/adjoints_ex6/element_postprocess.C:46:52: error: call of overloaded 'VectorValue(libMesh::Real, libMesh::Real)' is ambiguous
       NumberVectorValue U ((1-y)*(1-(2*x)),-x*(1-x));
                                                    ^
In file included from ../../../include/libmesh/fe_abstract.h:26:0,
                 from ../../../include/libmesh/fe_base.h:28,
                 from ../../../../examples/adjoints/adjoints_ex6/element_postprocess.C:4:
../../../include/libmesh/vector_value.h:205:1: note: candidate: libMesh::VectorValue<T>::VectorValue(const libMesh::TypeVector<double>&, const libMesh::TypeVector<double>&) [with T = std::complex<double>]
 VectorValue<T>::VectorValue (const TypeVector<Real> & p_re,
 ^~~~~~~~~~~~~~
../../../include/libmesh/vector_value.h:147:1: note: candidate: libMesh::VectorValue<T>::VectorValue(T, T, T) [with T = std::complex<double>]
 VectorValue<T>::VectorValue (const T x,
 ^~~~~~~~~~~~~~
```

**Comment by vikramvgarg** [comment]
@roystgnr @jwpeterson All checks passed. Ok to merge ?

File: examples/adjoints/adjoints_ex6/adjoints_ex6.C [hunk 1]
```diff
@@ -0,0 +1,526 @@
...
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+
+// <h1>Adjoints Example 6 - Poisson Equation in square domain with AdjointRefinementErrorEstimator</h1>
+// \author Vikram Garg
+// \date 2017
+//
+// This example solves the Poisson equation, whose solution displays a sharp layer,
+// with QoI based adjoint error estimation and adaptive mesh refinement. The exact
+// solution is . This example also illustrates the use of the adjoint Dirichlet boundary
```
Comment (roystgnr) [inline examples/adjoints/adjoints_ex6/adjoints_ex6.C]: "The exact solution is ."?

Comment (vikramvgarg) [inline examples/adjoints/adjoints_ex6/adjoints_ex6.C]: Its in poisson.in.

File: examples/adjoints/adjoints_ex6/adjoints_ex6.C [hunk 2]
```diff
@@ -0,0 +1,526 @@
...
+// condition capability, necesary for handling flux QoIs. Since we use the adjoint capabilities
+// of libMesh in this example, we use the DiffSystem framework. This file (adjoints_ex6.C)
```
Comment (roystgnr) [inline examples/adjoints/adjoints_ex6/adjoints_ex6.C]: This seems misleading - FEMSystem is the only thing with *all* our adjoints support, but we can do adjoint solves with any ImplicitSystem subclass, and this example doesn't actually call any unsupported-by-default code paths, does it?

File: examples/adjoints/adjoints_ex6/adjoints_ex6.C [hunk 3]
```diff
@@ -0,0 +1,526 @@
...
+// There is no need for element and side qoi derivative functions, since the adjoint RHS is supplied
+// by the adjoint dirichlet boundary condition.
+
+// WARNING: To maintain consistency, the flux must be computed internally in the library as R^h(u^h, L^h)
+// where R^h is the residual vector, u^h is the solution, and L^h the lift function at the current mesh.
+// It is w.r.t to this QoI evaluation that the adjoint error estimate R^h+([u^h]+, z^h+ - [L^h]+) is a
+// valid estimate. Automatic flux QoI evaluation via Adjoint Dirichlet boundary conditions is forthcoming
+// in libMesh, after which flux QoIs will be computed internally and there will be no need for
+// element_postprocess.C
+
```
Comment (roystgnr) [inline examples/adjoints/adjoints_ex6/adjoints_ex6.C]: Ah, wait, here it is.  FEMSystem is the only code that *automatically* handles adjoint DBC based flux QoI evaluation in assemble_qoi(), right?  Any *other* system types should need to handle things consistently, but that should be a completed feature in fem_system.C

File: examples/adjoints/adjoints_ex6/adjoints_ex6.C (hunk 1 above)
Comment (roystgnr) [inline examples/adjoints/adjoints_ex6/adjoints_ex6.C]: Fix that in the comment?

File: examples/adjoints/adjoints_ex6/adjoints_ex6.C (hunk 3 above)
Comment (vikramvgarg) [inline examples/adjoints/adjoints_ex6/adjoints_ex6.C]: Yes, but for stabilized QoIs even direct assemble_qoi will not be right. User will need to specify a residual evaluation physics, and as of now, only ARefEE will compute the correct flux.

File: examples/adjoints/adjoints_ex6/adjoints_ex6.C (hunk 1 above)
Comment (vikramvgarg) [inline examples/adjoints/adjoints_ex6/adjoints_ex6.C]: Of course.

Comment (vikramvgarg) [inline examples/adjoints/adjoints_ex6/adjoints_ex6.C]: Pushed.

File: examples/adjoints/adjoints_ex6/adjoints_ex6.C (hunk 3 above)
Comment (roystgnr) [inline examples/adjoints/adjoints_ex6/adjoints_ex6.C]: That's getting into more heinous detail than this unstabilized example needs, and it doesn't really excuse being misleading in the comment here, right?  "The flux we compute internally is suboptimal in the preasymptotic regime for some stabilized formulations" would have been correct, but "we don't compute fluxes internally" isn't, and the latter is how the comment here reads.

File: examples/adjoints/adjoints_ex6/adjoints_ex6.C (hunk 1 above)
Comment (roystgnr) [inline examples/adjoints/adjoints_ex6/adjoints_ex6.C]: Thanks!

**Comment by roystgnr** [comment]
AFK for tonight, but that comment is the last issue I have.  We should let @jwpeterson sign off before merging, though.

**Comment by jwpeterson** [comment]
I'm going to fetch this branch and reorganize it, fix commit metadata and formatting, etc. I'll either put up a new PR or merge when done.

**Comment by jwpeterson** [comment]
Merged to master in 129bc77.

**Pull request closed without merging**